toto/util.py

```diff
@@ -1,31 +1,108 @@
-"""
-    TODO: markup
-"""
 import os
 import sys
 import pickle
 import json
+import getpass
 
 import toto.ssl_crypto.keys
+import toto.ssl_crypto.formats
 import toto.log as log
 
-def create_and_persist_or_load_key(filename):
-  """Throw-away function that loads a key file or creates and
-  stores it if it does not expist.
 
-  Todo:
-      THROW AWAY!!!
+def generate_and_write_rsa_keypair(filepath, password):
   """
+  <Purpose>
+    Generate an RSA key file, create an encrypted PEM string (using 'password'
+    as the pass phrase), and store it in 'filepath'.  The public key portion of
+    the generated RSA key is stored in <'filepath'>.pub
 
-  if not os.path.isfile(filename):
-    log.warning("key '%s' was not found!" % filename)
-    log.doing("create new key '%s'" % filename)
-    key_dict = toto.ssl_crypto.keys.generate_rsa_key()
-    with open(filename, "w+") as fp:
-      key_data = toto.ssl_crypto.keys.format_keyval_to_metadata(
-          key_dict["keytype"], key_dict["keyval"], private=True)
-      fp.write(json.dumps(key_data))
-      return key_dict
+  <Arguments>
+    filepath:
+      The public and private key files are saved to <filepath>.pub, <filepath>
+      respectively
+
+    password:
+      The password used to encrypt 'filepath'
+
+  <Exceptions>
+    ssl_commons.FormatError, if the arguments are improperly formatted
+
+  <Side Effects>
+    Writes key files to '<filepath>' and '<filepath>.pub'
+
+  <Returns>
+    None.
+  """
+
+  toto.ssl_crypto.formats.PATH_SCHEMA.check_match(filepath)
+  toto.ssl_crypto.formats.PASSWORD_SCHEMA.check_match(password)
+
+  rsa_key = toto.ssl_crypto.keys.generate_rsa_key()
+
+  public = rsa_key['keyval']['public']
+  private = rsa_key['keyval']['private']
+  private_pem = toto.ssl_crypto.keys.create_rsa_encrypted_pem(private, password)
+
+  with open(filepath + '.pub', 'w') as fo_public, \
+      open(filepath, 'w') as fo_private:
+
+    fo_public.write(public.encode('utf-8'))
+    fo_private.write(private_pem.encode('utf-8'))
```

Comment:
Hm. I think I misused the `with` statement. Actually I intended something like a transaction, i.e. write both files or none. But this is not what `with` does.
I can definitely use two `with` blocks.